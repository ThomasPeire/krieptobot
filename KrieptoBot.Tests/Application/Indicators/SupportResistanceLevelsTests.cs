using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using KrieptoBot.Application.Indicators;
using KrieptoBot.DataVisualizer;
using KrieptoBot.DataVisualizer.Extensions;
using KrieptoBot.Domain.Trading.ValueObjects;
using KrieptoBot.Infrastructure.Bitvavo.Dtos;
using KrieptoBot.Infrastructure.Bitvavo.Extensions;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using NUnit.Framework;
using Plotly.NET;
using Snapshooter.NUnit;

namespace KrieptoBot.Tests.Application.Indicators;

internal class SupportResistanceLevelsTests
{
    private IEnumerable<Candle> _candles;

    [SetUp]
    public void Setup()
    {
        InitCandles();
    }

    private void InitCandles()
    {
        var candlesJson = File.ReadAllText(@"./MockData/Bitvavo/candles_btc-eur_4h.json");
        var deserializedCandles = JsonConvert.DeserializeObject(candlesJson) as JArray;
        _candles = deserializedCandles.Select(x =>
            new CandleDto
            {
                TimeStamp = x.Value<long>(0),
                Open = x.Value<decimal>(1),
                High = x.Value<decimal>(2),
                Low = x.Value<decimal>(3),
                Close = x.Value<decimal>(4),
                Volume = x.Value<decimal>(5)
            }.ConvertToKrieptoBotModel());
    }

    [Test]
    public void CalculateWithStrength_Should_DetermineSupportAndResistanceLevelsWithStrength()
    {
        var datetimeFrom = new DateTime(2021, 11, 01);
        var dateTimeTo = new DateTime(2021, 11, 30);

        var candlesToWorkWith = _candles
            .Where(x => x.TimeStamp >= datetimeFrom && x.TimeStamp <= dateTimeTo)
            .OrderBy(x => x.TimeStamp).ToArray();

        var supportLevels = new SupportResistanceLevels().Calculate(candlesToWorkWith).ToArray();
        Snapshot.Match(supportLevels);

#if DEBUG
        var candleVisualizer = new CandlesVisualizer();
        var candleChart = candleVisualizer.Visualize(candlesToWorkWith);
        candleChart = candleChart.AddSupportLevels(supportLevels, datetimeFrom, dateTimeTo);
        candleChart.WithSize(1920, 1080).WithConfig(Config.init(Responsive: true)).Show();
#endif
    }
}