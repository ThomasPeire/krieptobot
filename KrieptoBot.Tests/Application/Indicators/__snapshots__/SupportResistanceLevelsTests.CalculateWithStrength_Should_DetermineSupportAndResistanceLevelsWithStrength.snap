[
  {
    "Value": {
      "Value": 54000.0
    },
    "From": "2021-11-01T08:00:00Z"
  },
  {
    "Value": {
      "Value": 52090.0
    },
    "From": "2021-11-01T16:00:00Z"
  },
  {
    "Value": {
      "Value": 55500.0
    },
    "From": "2021-11-02T16:00:00Z"
  },
  {
    "Value": {
      "Value": 59080.0
    },
    "From": "2021-11-09T00:00:00Z"
  },
  {
    "Value": {
      "Value": 59688.0
    },
    "From": "2021-11-10T12:00:00Z"
  },
  {
    "Value": {
      "Value": 57290.0
    },
    "From": "2021-11-11T08:00:00Z"
  },
  {
    "Value": {
      "Value": 49042.0
    },
    "From": "2021-11-19T00:00:00Z"
  },
  {
    "Value": {
      "Value": 53197.0
    },
    "From": "2021-11-20T20:00:00Z"
  },
  {
    "Value": {
      "Value": 46000.0
    },
    "From": "2021-11-26T08:00:00Z"
  },
  {
    "Value": {
      "Value": 47538.0
    },
    "From": "2021-11-28T00:00:00Z"
  }
]
